File: Plugins/TrayPerfmon.Plugin.SampleCs/SamplePlugin.cs
using System;
using System.ComponentModel.Plugin;
using System.Diagnostics;
using System.Drawing;
using System.Drawing.Drawing2D;
using System.Linq;

namespace TrayPerfmon.Plugin.SampleCs
{
    // Export plugin description.
    [Plugin("SamplePluginCs", "Sample Plugin: % Disk Read/Write Time")]
    public class SamplePlugin : NotifyIconPlugin
    {
        /// <summary>
        /// Construct sample 15 frame par second.
        /// </summary>
        public SamplePlugin() : base(1000 / 15) {
        }

        /// <summary>
        /// Apply settings.
        /// </summary>
        protected override void ApplySettings() {
            foreach (var counter in _performanceCounter) {
                counter.Dispose();
            }
            _performanceCounter = [
                new PerformanceCounter("PhysicalDisk", "% Disk Read Time", "_Total", true),
                new PerformanceCounter("PhysicalDisk", "% Disk Write Time", "_Total", true),
            ];
            _brushes = new Brush[] { Brushes.Cyan, Brushes.Magenta };
        }

        protected override void Clear(Graphics graphics) {
            graphics.Clear(Color.Transparent);
        }

        /// <summary>
        /// Override draw method.
        /// </summary>
        protected override void Draw(Graphics graphics) {
            var value = _performanceCounter.Select(counter => counter.NextValue()).ToArray();
            graphics.SmoothingMode = SmoothingMode.AntiAlias;
            var bounds = graphics.VisibleClipBounds;
            // Draw Read/Write % value
            const int row = 2, column = 4;
            const int count = row * column, sample = 100 / count;
            float x = bounds.X, y = bounds.Y + bounds.Height / 2f;
            float width = bounds.Width / column, height = bounds.Height / 2f / row;
            for (var rw = 0; rw < 2; ++rw) {
                var next = (int)(value[rw] + sample - .1f) / sample;
                for (var i = 0; i < count; ++i) {
                    var brush = i < next ? _brushes[rw] : new SolidBrush(Color.FromArgb(0xC0, 0x00, 0x00, 0x00));
                    var dx = i % column;
                    var dy = i / column;
                    graphics.FillRectangle(brush, x + width * dx, y * rw + height * dy, width, height);
                }
            }
        }

        protected override void Dispose(bool disposing) {
            if (disposing) {
                foreach (var counter in _performanceCounter) {
                    counter.Dispose();
                }
            }
            base.Dispose(disposing);
        }

        PerformanceCounter[] _performanceCounter = [];

        /// <summary>Draw brush for disk access.</summary>
        Brush[] _brushes;
    }
}

File: Plugins/TrayPerfmon.Plugin.SampleVb/SamplePlugin.vb
Imports System.ComponentModel.Plugin
Imports System.Drawing
Imports System.Drawing.Drawing2D

<Plugin("SamplePluginVb", "Sample Plugin: Memory and Paging File")>
Public Class SamplePlugin
    Inherits NotifyIconPlugin

    Protected Overrides ReadOnly Property Factories As Lazy(Of PerformanceCounter)()
        Get
            Return _factories
        End Get
    End Property

    Sub New()
        MyBase.New(2, 1000 / 15)
    End Sub

    Protected Overrides Sub Clear(graphics As Graphics)
        ' Clear icon image
        graphics.Clear(Color.Transparent)
    End Sub

    Protected Overrides Sub Draw(graphics As Graphics, value() As Single)
        graphics.SmoothingMode = SmoothingMode.AntiAlias
        Dim bounds = graphics.VisibleClipBounds
        ' Draw Memory and Paging File
        Const row = 2, column = 4
        Const count = row * column, sample = 100 / count
        Dim x = bounds.X, y = bounds.Y + bounds.Height / 2.0F
        Dim width = bounds.Width / column, height = bounds.Height / 2.0F / row
        For rw = 0 To 1
            Dim [next] = CInt(value(rw) + sample - 0.1F) / sample
            For i = 0 To count
                Dim brush = IIf(i < [next], _brushes(rw), Brushes.Black)
                Dim dx = i Mod column
                Dim dy = i \ column
                graphics.FillRectangle(brush, x + width * dx, y * rw + height * dy, width, height)
            Next
        Next
    End Sub

    Shared Sub New()
        ' Open 'Server Explorer' window and find your own performance counter in local machine
        _factories = New Lazy(Of PerformanceCounter)() {
            New Lazy(Of PerformanceCounter)(Function() New PerformanceCounter("Memory", "% Committed Bytes In Use", True)),
            New Lazy(Of PerformanceCounter)(Function() New PerformanceCounter("Paging File", "% Usage", "_Total", True))
        }
        ' Create Read/Write color brush
        _brushes = New Brush() {Brushes.Cyan, Brushes.Magenta}

    End Sub

    Shared _factories As Lazy(Of PerformanceCounter)()
    Shared _brushes As Brush()

End Class
